ENH Change fit_transform of MissingIndicator class to get mask only once

harsh020 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes #14278


#### What does this implement/fix? Explain your changes.
The implementation fixes the function _get_missing_features_info from being called twice in fit_transform and also adds a test to check the result are correct.

--Declare a private instance variable (double leading underscore variable/mangling) in fit function, to store the return values of _get_missing_features_info.
-- Called fit from fit_transform.
--Used the private instance variable, when implementing transform in fit_transform to retrieve data.

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Commit:
Change fit_transform of MissingInformation class to prevent it from calling _get_missing_features_info twice (#14278)

Commit:
Remove whitespaces

Commit:
Modify comments in MissingIndicator for readability

Commit:
Modify comment in MissingIndicator

Commit:
Add entire fit to fit fit_transform to prevent creating an extra data member

amueller commented on sklearn/impute/_base.py:
```diff
@@ -667,7 +663,31 @@ def fit_transform(self, X, y=None):
+        self._n_features = X.shape[1]
+
+        if self.features not in ('missing-only', 'all'):
+            raise ValueError("'features' has to be either 'missing-only' or "
```
You shouldn't duplicate the code ideally. It's not covered by the tests which is why you're seeing the code coverage failure.

harsh020 replied:
Then what should I do, because I need to do the exact same things, I can't change the variable name (can I??). I can't call fit function because if I do I need to create a data member to store what the function returns. 
Any help??

Thanks.

harsh020 replied:
@amueller Then what should I do, because I need to do the exact same things, I can't change the variable name (can I??). I can't call fit function because if I do I need to create a data member to store what the function returns which raises more failures during check because the tests do not know have any information about it. 
Any help??

Thanks.

harsh020 replied:
@amuller Can I use mangling??

amueller commented on sklearn/impute/_base.py:
```diff
@@ -1,7 +1,3 @@
-# Authors: Nicolas Tresegnie <nicolas.tresegnie@gmail.com>
```
Why did you remove the author information?

harsh020 replied:
I didn't do that intentionally, its a mistake, sorry I'll restore it immediately.

harsh020 replied:
Wait a minute, its still there in my local copy.

harsh020 replied:
See,
![Screen Shot 2019-07-14 at 10 22 34 PM](https://user-images.githubusercontent.com/39561084/61186702-f2731400-a685-11e9-816d-bec511652877.png)

I don't know why the diff shows so.

Commit:
Use mangling to rsolve codecov/patch

Commit:
Restore author information

amueller commented on sklearn/impute/_base.py:
```diff
@@ -612,7 +612,8 @@ def fit(self, X, y=None):
                              "Got {!r} instead.".format(self.sparse))
 
-        self.features_ = self._get_missing_features_info(X)[1]
+        self.__missing_features_info = self._get_missing_features_info(X)
```
I think we usually only use a single underscore, but otherwise this solution seems good to me.
Do we have a test that checks that ``fit(X).transform(X) == fit_transform(X)``?

harsh020 replied:
A single underscore will declare a protected instance variable, which can be accessed outside the class making data transparent, hence I've used double underscore which can't be accessed outside the class which would maintain opacity as the user need not know that data, he is already getting it when using fit_transform.
And moreover I think that can raise issue fit PR checks as the fit function returns self, which would include that protected member, but the tests do not know that fit does return it, so we need to add that to tests.
So, is it okay using private instance variable or should I change it to single?

harsh020 replied:
> Do we have a test that checks that `fit(X).transform(X) == fit_transform(X)`?
I don't think so. I'm sorry but I'm very new to test stuff, but I had a look in the test file and don't think that we have something that checks it.

amueller replied:
could you please add one? it should be quite simple.

harsh020 replied:
should I add it to the pre existing test function test_missing_indicator_new

harsh020 replied:
okay, I did add a test, I'm pushing it. Please do let me know if any changes are to be made

Commit:
Add fit_transform test to MissingIndicator

Commit:
Add newline indents to fit_transform test

Commit:
Correctly indent fit_transform test

Commit:
Fix indentation of fit_transform test

harsh020 commented:
@amueller I did add a test, please have a look and let me know if I need to make some changes.

thomasjpfan commented:
Storing `missing_features_info`, which includes the entire mask seems a little heavy for the `MissingIndicator`. An alternative is to move all of `fit` into a private method `_fit` and have `_fit` return `missing_features_info`. 

1. `fit` will call `_fit` and return self.

2. `fit_transform` will call `_fit` and use the returned `missing_features_info` to perform the transform.

harsh020 commented:
@thomasjpfan I'm a bit confused please help out here - 
So, what you are saying is I should make a protected function `_fit` and move entire `fit` to it and store and return `missing_features_info`. Then I should call` _fit` from fit and return `self` and also call `_fit` from `fit_transform`, to get that stored data.
How is it different from my approach?? I'll still be storing return values in `missing_features_info`.

thomasjpfan commented:
`_fit` will not store `missing_features_info` it will only return it.

harsh020 commented:
Okay I see, where you are going. Amazing, thats a nice idea! Thanks!

Commit:
Memory optimization for MissingIndicator efficiency fix

Commit:
Add blankline to pass linter check

harsh020 commented:
Just out of curiocity, what if I would have stored the return values in a local variables in the` _fit` function, would it have had the lifespan only when we called the `_fit` function or until the object of `MissingIndicator` was destroyed? 
And second thing, why do we use protected function (with a single _) rather than a private function (with a double _)?

thomasjpfan commented:
> would it (the local variable having the return values) have had the lifespan only when we called the _fit function

It will only exist in `_fit`

> And second thing, why do we use protected function (with a single _) rather than a private function (with a double _)?

In scikit-learn we use a single `_` to mean "non-public". Currently, we do not use the double underscore.

harsh020 commented:
@thomasjpfan Thanks for clearing that up, I made the necessary changes and pushed them. Can you please have a look and let me know. I was also asked to add a test for the same, can you please have a look at it too (bcs I'm very new to tests).
If every thing seems fine whom should I request to merge the branch?

Thanks.

thomasjpfan commented on sklearn/impute/_base.py:
```diff
@@ -612,7 +612,23 @@ def fit(self, X, y=None):
                              "Got {!r} instead.".format(self.sparse))
 
-        self.features_ = self._get_missing_features_info(X)[1]
+        return self._get_missing_features_info(X)
```
So that `_fit` actually fits and learns something:

```py
missing_features_info = self._get_missing_features_info(X)
self.features_ = missing_features_info[1]
return missing_features_info
```

thomasjpfan commented on sklearn/impute/_base.py:
```diff
@@ -612,7 +612,23 @@ def fit(self, X, y=None):
+        self : object
+            Returns self.
+        """
+        self.features_ = self._fit(X, y)[1]
```
With the change suggested in `_fit` this can be:

```suggestion
        self._fit(X, y)
```

thomasjpfan commented on sklearn/impute/_base.py:
```diff
@@ -667,7 +683,14 @@ def fit_transform(self, X, y=None):
         """
-        return self.fit(X, y).transform(X)
+        imputer_mask, features = self._fit(X, y)
+        self.features_ = features
```
`self.features_ = features` can be removed with the suggested change in `_fit`.

thomasjpfan commented:
> If every thing seems fine whom should I request to merge the branch?

After I approve this PR, we need to wait for another core dev to review the PR before it is merged.

harsh020 commented:
@thomasjpfan Yup, I also thought to do so, but since you asked me not to store in missing_features_info, I thought you don't want me to store it in a local variable too. So I didn't  do it. Anyways, I patch it up.

Commit:
Store return values in missing_features_info

Commit:
Remove blankline

harsh020 commented:
@thomasjpfan Hi, I don't mean to disturb you but, is every thing fine or do I need to make any changes in this PR?

thomasjpfan reviewed (approved):
LGTM

amueller commented on sklearn/impute/_base.py:
```diff
@@ -612,7 +612,26 @@ def fit(self, X, y=None):
+        missing_features_info = self._get_missing_features_info(X)
+        self.features_ = missing_features_info[1]
+
+        return missing_features_info
```
Why not just return the mask if ``features_`` is already stored as an attribute?

harsh020 replied:
@amueller Good point, I'll do it right now.
Thanks.

Commit:
Return only mask in _fit

jnothman commented on sklearn/impute/_base.py:
```diff
@@ -667,7 +686,13 @@ def fit_transform(self, X, y=None):
-        return self.fit(X, y).transform(X)
+        imputer_mask = self._fit(X, y)
+
+        if (self.features == "missing-only") and \
```
these parentheses are unnecessary

jnothman commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -1008,6 +1008,13 @@ def test_missing_indicator_new(missing_values, arr_type, dtype, param_features,
+                                         sparse=False)
+    fit_transformer = indicator_test_ft.fit(X_fit).transform(X_fit)
+    fit_transformer_ = indicator_test_ft.fit_transform(X_fit)
+    assert_array_equal(fit_transformer, fit_transformer_)
```
I'm pretty sure this is already tested. If this change is improving efficiency, and not changing functionality, it does not require a new test.

harsh020 replied:
Ok, I'll remove it, I added it because I was asked to. And I think I don't see where it is being checked. Please correct me if I'm wrong.

Thanks.

jnothman replied:
I think this is meant to be checked within check_estimator

harsh020 replied:
Ok, so I'll remove it.

Thanks.

jnothman commented on sklearn/impute/_base.py:
```diff
@@ -667,7 +686,13 @@ def fit_transform(self, X, y=None):
+        imputer_mask = self._fit(X, y)
+
+        if (self.features == "missing-only") and \
+                (self.features_.size < self._n_features):
```
I think this is a sufficient condition. Remove the other.

harsh020 replied:
You mean `(self.features_.size < self._n_features)`, this one?

jnothman replied:
I mean remove `(self.features == "missing-only")`

harsh020 replied:
But what if self.features is 'all' bcs in _fit we are ensuring that the features should be either 'missing-only' or 'all'. So, this particular block of code will execute even when the value of self.features is 'all'.
Is this acceptable?? will it not create any further problem??

Thanks.

jnothman replied:
If features='all', then `self.features_.size <  self._n_features` can not be true.

"mask" is, to my understanding, taken from painting where the mask covers some parts of the canvas and uncovers others. It's used in numeric computing to mean a Boolean array used to select cells from an array of the same size.

harsh020 replied:
> If features='all', then `self.features_.size < self._n_features` can not be true.

oh yeah, sorry, for that silly thing.

I'll make the necessary changes right now.

Thanks

jnothman reviewed (commented):
Please add an Efficiency entry to the change log at `doc/whats_new/v0.22.rst`. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`

harsh020 commented:
> Please add an Efficiency entry to the change log at `doc/whats_new/v0.22.rst`. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`

@jnothman Can I include the names of all mentors that helped me out?

jnothman commented:
> @jnothman Can I include the names of all mentors that helped me out?

Unless we've made substantive contributions it would be unconventional to do so. The record of mentoring is here on the issue tracker, and the core developers do not tend to seek further recognition for small patches.

harsh020 commented:
> I'm pretty sure this is already tested.

@jnothman Ok, I'll remove it, I added it because I was asked to. And I think I don't see where it is being checked.

> I think this is a sufficient condition. Remove the other.

You mean remove, `(self.features_.size < self._n_features)`, this one? I don't think that any of those conditions are being checked in `_fit`.

Please correct me if I'm wrong.

Thanks.

harsh020 commented:
@jnothman Sorry, I am creating a mess here. Please help me out here. As I was getting merge conflict, by mistake I merged my missing_indicator branch to upstream/master branch, to correct it  I merged the upstream/master changes to local master and reset my missing_indicator head and force pushed it. Nut still I'm getting merge conflicts.
Should I delete the branch, do the necessary changes and then push it.
I know I'm creating a huge mess (my first PR), Sorry. Please help me out.

Thanks.

Commit:
Add entry of efficiency and remove test and unnecessary condition

harsh020 commented:
@jnothman I think I reversed back - I restored my local repo back where I didn't merge (commit ffe3a5a), and also did the necessary changes (removed the unnecessary condition, removed tests and added efficiency entry), I pushed it and think every thing just got back where the merge conflict arose. Should I save the conflict by keeping both the logs or removing the log which was merged in master.

Commit:
Merge branch 'master' into missing_indicator

thomasjpfan commented:
> Should I save the conflict by keeping both the logs or removing the log which was merged in master.

We prefer if you do not do force pushes to remove commits.

thomasjpfan commented on sklearn/impute/_base.py:
```diff
@@ -667,7 +686,12 @@ def fit_transform(self, X, y=None):
-        return self.fit(X, y).transform(X)
+        imputer_mask = self._fit(X, y)
+
+        if (self.features_.size < self._n_features):
```
```suggestion
        if self.features_.size < self._n_features:
```

jnothman commented:
Well done for fixing it up though

jnothman commented on sklearn/impute/_base.py:
```diff
@@ -598,7 +598,7 @@ def fit(self, X, y=None):
         Returns
         -------
         self : object
```
This is no longer accurate

harsh020 replied:
Do you mean the return comment of `_fit` or `fit`?

Commit:
Fix return comment for _fit function of MissingIndicator class

Commit:
Remove unnecessary parenthesis in fit_tranform function in MissingIndicator class

harsh020 commented:
> We prefer if you do not do force pushes to remove commits.

@thomasjpfan Sorry,  just got scared for messing things up. Won't happen again.

harsh020 commented:
> Well done for fixing it up though

@jnothman Thanks! Please have a look and let me know if I did it correctly - 
(Visited my course, and revised how to resolve merge conflicts)
1. Pulled upstream/master to my local master.
2. Switched to missing_indicator branch and merged master to it.
3. Resolved merge conflicts.
4. Pushed missing_indicator branch to origin.

> This is no longer accurate

Are you talking about the return comment of `_fit`? If yes, I fixed it. If no, sorry, I don't follow, please tell me what is not accurate, the comments of fit?

Thanks.

thomasjpfan commented on sklearn/impute/_base.py:
```diff
@@ -597,8 +597,10 @@ def fit(self, X, y=None):
-            Returns self.
+        Xt : {ndarray or sparse matrix}, shape (n_samples, n_features)
+            The missing indicator for input data. The data type of ``Xt``
+            will be boolean.
```
You can take use the docstring from `_get_missing_features_info`, since `_fit` returns one of the outputs of `_get_missing_features_info`:

```py
        imputer_mask : {ndarray or sparse matrix}, shape \
(n_samples, n_features) or (n_samples, n_features_with_missing)
            The imputer mask of the original data.
```

harsh020 replied:
Sure. 

Thanks.

thomasjpfan replied:
Yup I agree with just `(n_samples, n_features)`.

thomasjpfan commented:
>  Sorry, just got scared for messing things up. Won't happen again.

@harsh020 It is okay! Thank you for working on this PR! 😄

harsh020 commented:
> It is okay! Thank you for working on this PR! 😄

The pleasure was all mine. Thank you for helping me all along, as without you guys it would have been impossible for me to work on this issue.

Thanks.

Commit:
Fix return docstring of _fit of MissingIndicator class

thomasjpfan commented on sklearn/impute/_base.py:
```diff
@@ -597,8 +597,10 @@ def fit(self, X, y=None):
         -------
-        self : object
-            Returns self.
+        imputer_mask : {ndarray or sparse matrix},
```
```py
        imputer_mask : {ndarray or sparse matrix}, shape (n_samples, \
        n_features)
            The imputer mask of the original data.
```

harsh020 replied:
Is something wrong?

thomasjpfan replied:
The spacing. This is only for consistency. This docstring is not going to be rendered since `_fit` is private.

Notice the use of `\` and how `n_features)` is aligned and how many white spaces there are before `The imputer...`.

harsh020 replied:
Done.

Thanks.

Commit:
Fix spacing in docstring in _fit in MissingIndicator

jnothman commented:
Thank you @harsh020!

harsh020 commented:
> Thank you @harsh020!

@jnothman The pleasure was all mine. Thanks to you guys, would not have been possible without you guys.

harsh020 commented:
@jnothman Can I delete my branch?

jnothman commented:
You are welcome to.

jnothman commented on doc/whats_new/v0.22.rst:
```diff
@@ -107,6 +107,11 @@ Changelog
+- |Efficiency| :func:`ensemble.MissingIndicator.fit_transform` the
+  _get_missing_features_info function is now called once when calling
+  fit_transform for MissingIndicator class. :pr:`14356` by :user:
+  `Harsh Soni <harsh020>`.
```
This doesn't render correctly with space before it. It must be right after the `:`

harsh020 replied:
@jnothman I did open a PR, kindly review it, https://github.com/scikit-learn/scikit-learn/pull/15027

jnothman reviewed (commented):
Please open a new pr fixing this up, @harsh020. Sorry to have missed it

harsh020 commented on doc/whats_new/v0.22.rst:
```diff
@@ -107,6 +107,11 @@ Changelog
   preserve the class balance of the original training set. :pr:`14194`
   by :user:`Johann Faouzi <johannfaouzi>`.
 
+- |Efficiency| :func:`ensemble.MissingIndicator.fit_transform` the
```
oh, my bad. Really sorry for the blunder. I'm immediately opening a new pr and fixing it up.
